File: web/e-platform-web/src/services/TransactionService.js
const API_URL = process.env.REACT_APP_BFF_API_URL || 'https://tunsiaed.onrender.com';

class TransactionService {
  /**
   * Creates a new transaction
   * @param {Object} transactionData - Transaction data
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Created transaction data
   */
  static async createTransaction(transactionData, token) {
    const res = await fetch(`${API_URL}/api/v1/transaction`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(transactionData),
    });
    if (!res.ok) throw new Error('Failed to create transaction');
    return res.json();
  }

  /**
   * Fetches transactions for the authenticated user
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of user's transactions
   */
  static async getUserTransactions(token) {
    const res = await fetch(`${API_URL}/api/v1/transaction/my-transactions`, {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
    if (!res.ok) throw new Error('Failed to fetch user transactions');
    return res.json();
  }

  /**
   * Fetches a specific transaction by ID
   * @param {string} transactionId - The ID of the transaction
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Transaction data
   */
  static async getTransactionById(transactionId, token) {
    const res = await fetch(`${API_URL}/api/v1/transaction/${transactionId}`, {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
    if (!res.ok) throw new Error('Failed to fetch transaction');
    return res.json();
  }

  /**
   * Fetches transactions associated with a payment
   * @param {string} paymentId - The ID of the payment
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of transactions for the payment
   */
  static async getTransactionsByPayment(paymentId, token) {
    const res = await fetch(`${API_URL}/api/v1/transaction/payment/${paymentId}`, {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
    if (!res.ok) throw new Error('Failed to fetch transactions by payment');
    return res.json();
  }

  /**
   * Fetches transactions for a specific course
   * @param {string} courseId - The ID of the course
   * @param {string} token - Authentication token (requires admin or instructor role)
   * @returns {Promise<Array>} List of transactions for the course
   */
  static async getCourseTransactions(courseId, token) {
    const res = await fetch(`${API_URL}/api/v1/transaction/course/${courseId}`, {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
    if (!res.ok) throw new Error('Failed to fetch course transactions');
    return res.json();
  }

  /**
   * Fetches transactions by status
   * @param {string} status - Transaction status (e.g., 'pending', 'completed')
   * @returns {Promise<Array>} List of transactions with the specified status
   */
  static async getTransactionsByStatus(status,token) {
    const res = await fetch(`${API_URL}/api/v1/transaction/status/${status}`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
    if (!res.ok) throw new Error('Failed to fetch transactions by status');
    return res.json();
  }
  /**
   * Updates an existing transaction
   * @param {string} transactionId - The ID of the transaction to update
   * @param {Object} transactionData - Updated transaction data
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Updated transaction data
   */
  static async updateTransaction(transactionId, transactionData, token) {
    const res = await fetch(`${API_URL}/api/v1/transaction/${transactionId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(transactionData),
    });
    if (!res.ok) throw new Error('Failed to update transaction');
    return res.json();
  }
}

export default TransactionService;

File: mobile/e_platform_mobile/lib/Service/transaction/transaction_service.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_endpoints.dart';

/// Transaction Service
/// Handles all transaction-related API calls
class TransactionService {
  final ApiClient _apiClient = ApiClient();

  /// Get all transactions for current user
  Future<List<Map<String, dynamic>>> getTransactions({
    int? page,
    int? limit,
  }) async {
    final response = await _apiClient.get(
      ApiEndpoints.transactions,
      queryParameters: {
        if (page != null) 'page': page,
        if (limit != null) 'limit': limit,
      },
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get transaction by ID
  Future<Map<String, dynamic>> getTransactionById(
    String transactionId,
  ) async {
    final response = await _apiClient.get(
      ApiEndpoints.transactionById(transactionId),
    );
    return response.data;
  }

  /// Get transactions by status
  Future<List<Map<String, dynamic>>> getTransactionsByStatus(
    String status,
  ) async {
    final response = await _apiClient.get(
      ApiEndpoints.transactions,
      queryParameters: {'status': status},
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get transaction history
  Future<List<Map<String, dynamic>>> getTransactionHistory({
    DateTime? startDate,
    DateTime? endDate,
  }) async {
    final response = await _apiClient.get(
      '${ApiEndpoints.transactions}/history',
      queryParameters: {
        if (startDate != null) 'startDate': startDate.toIso8601String(),
        if (endDate != null) 'endDate': endDate.toIso8601String(),
      },
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Create transaction
  Future<Map<String, dynamic>> createTransaction(
    Map<String, dynamic> transactionData,
  ) async {
    final response = await _apiClient.post(
      ApiEndpoints.transactions,
      data: transactionData,
    );
    return response.data;
  }

  /// Get transactions by payment
  Future<List<Map<String, dynamic>>> getTransactionsByPayment(
    String paymentId,
  ) async {
    final response = await _apiClient.get(
      '${ApiEndpoints.transactions}/payment/$paymentId',
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get course transactions
  Future<List<Map<String, dynamic>>> getCourseTransactions(
    String courseId,
  ) async {
    final response = await _apiClient.get(
      '${ApiEndpoints.transactions}/course/$courseId',
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Update transaction
  Future<Map<String, dynamic>> updateTransaction(
    String transactionId,
    Map<String, dynamic> transactionData,
  ) async {
    final response = await _apiClient.put(
      ApiEndpoints.transactionById(transactionId),
      data: transactionData,
    );
    return response.data;
  }
}
